src/main/java/com/iexec/core/replicate/ReplicateSupplyService.java

```diff
@@ -0,0 +1,336 @@
+        }
+
+        if (task.getCurrentStatus().equals(TaskStatus.CONSENSUS_REACHED)) {
+            if (!replicate.containsContributedStatus())
+                return RecoveryAction.ABORT_CONSENSUS_REACHED;
+        }
+
+        RecoveryAction recoveryAction = null;
+
+        if (task.inRevealPhase()) {
+            recoveryAction = recoverReplicateInRevealPhase(task, replicate, chainReceipt);
+        }
+
+        if (task.inResultUploadPhase()) {
+            recoveryAction = recoverReplicateInResultUploadPhase(task, replicate);
+        }
+
+        if (task.inCompletionPhase()) {
+            return recoverReplicateIfRevealed(replicate);
+        }
+
+        return recoveryAction;
+    }
+
+    /**
+     * CREATED, ..., CAN_CONTRIBUTE         => RecoveryAction.CONTRIBUTE
+     * CONTRIBUTING + !onChain              => RecoveryAction.CONTRIBUTE
+     * CONTRIBUTING + done onChain          => updateStatus to CONTRIBUTED & go to next case
+     * CONTRIBUTED + !CONSENSUS_REACHED     => RecoveryAction.WAIT
+     * CONTRIBUTED + CONSENSUS_REACHED      => RecoveryAction.REVEAL
+     */
+
+    private RecoveryAction recoverReplicateInContributionPhase(Task task, Replicate replicate, ChainReceipt chainReceipt) {
+        String chainTaskId = task.getChainTaskId();
+        String walletAddress = replicate.getWalletAddress();
+
+        boolean beforeContributing = replicate.isBeforeStatus(ReplicateStatus.CONTRIBUTING);
+        boolean didReplicateStartContributing = replicate.getLastRelevantStatus().equals(ReplicateStatus.CONTRIBUTING);
+        boolean didReplicateContributeOnChain = replicatesService.didReplicateContributeOnchain(chainTaskId, walletAddress);
+
+        if (beforeContributing) {
+            return RecoveryAction.CONTRIBUTE;
+        }
+
+        if (didReplicateStartContributing && !didReplicateContributeOnChain) {
+            return RecoveryAction.CONTRIBUTE;
+        }
+
+        if (didReplicateStartContributing && didReplicateContributeOnChain) {
+
+            replicatesService.updateReplicateStatus(chainTaskId, walletAddress,
+                    ReplicateStatus.CONTRIBUTED, ReplicateStatusModifier.POOL_MANAGER,
+                    chainReceipt, "");
+        }
+
+        // we read the replicate from db to consider the changes added in the previous case
+        Optional<Replicate> oReplicateWithLatestChanges = replicatesService.getReplicate(chainTaskId, walletAddress);
+        if (!oReplicateWithLatestChanges.isPresent()) return null;
+
+        boolean didReplicateContribute = oReplicateWithLatestChanges.get()
+                .getLastRelevantStatus()
+                .equals(ReplicateStatus.CONTRIBUTED);
+
+        if (didReplicateContribute) {
+
+            if (!taskService.isConsensusReached(task)) {
+                return RecoveryAction.WAIT;
+            }
+
+            taskExecutorEngine.updateTask(chainTaskId);
+            return RecoveryAction.REVEAL;
+        }
+
+        return null;
```

Comment:
Let's return an optional rather than a null value. This is valid for all recoverReplicate* methods.

Reply:
with pleasure